deepchem/models/torch_models/gan.py

```diff
@@ -0,0 +1,758 @@
+        -------
+        A Tensor equal to the loss function to use for optimizing the generator.
+        """
+        return -torch.mean(torch.log(discrim_output + 1e-10))
+
+    def create_discriminator_loss(self, discrim_output_train,
+                                  discrim_output_gen):
+        """Create the loss function for the discriminator.
+
+        The default implementation is appropriate for most cases.  Subclasses can
+        override this if the need to customize it.
+
+        Parameters
+        ----------
+        discrim_output_train: Tensor
+            the output from the discriminator on a batch of training data.  This is
+            its estimate of the probability that each sample is training data.
+        discrim_output_gen: Tensor
+            the output from the discriminator on a batch of generated data.  This is
+            its estimate of the probability that each sample is training data.
+
+        Returns
+        -------
+        A Tensor equal to the loss function to use for optimizing the discriminator.
+        """
+        return -torch.mean(
+            torch.log(discrim_output_train + 1e-10) +
+            torch.log(1 - discrim_output_gen + 1e-10))
+
+    def discrim_loss_fn_wrapper(self, outputs, labels, weights):
+        """Wrapper around create_discriminator_loss for use with fit_generator.
+
+        Parameters
+        ----------
+        outputs: Tensor
+            the output from the discriminator on a batch of training data.  This is
+            its estimate of the probability that each sample is training data.
+        labels: Tensor
+            the labels for the batch.  These are ignored.
+        weights: Tensor
+            the weights for the batch.  These are ignored.
+
+        Returns
+        -------
+        the value of the discriminator loss function for this input.
+        """
+        discrim_output_train, discrim_output_gen = outputs
+        return self.create_discriminator_loss(discrim_output_train,
+                                              discrim_output_gen)
+
+    def gen_loss_fn_wrapper(self, outputs, labels, weights):
+        """Wrapper around create_generator_loss for use with fit_generator.
+
+        Parameters
+        ----------
+        outputs: Tensor
+            the output from the discriminator on a batch of generated data.  This is
+            its estimate of the probability that each sample is training data.
+        labels: Tensor
+            the labels for the batch.  These are ignored.
+        weights: Tensor
+            the weights for the batch.  These are ignored.
+
+        Returns
+        -------
+        the value of the generator loss function for this input.
+        """
+        discrim_output_train, discrim_output_gen = outputs
+        return self.create_generator_loss(discrim_output_gen)
+
+
+def _list_or_tensor(inputs):
+    if len(inputs) == 1:
+        return inputs[0]
+    return inputs
+
+
+class GANModel(TorchModel):
```

Comment:
Separate this into 2 PRs, the first adding the GAN, the second adding the GANModel